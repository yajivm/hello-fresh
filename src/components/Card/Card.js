import PropTypes from 'prop-types';
import { DeleteIc, EditIc  } from '../Icons';
import { CardWrapper, CardImage, CardContent, CardTitle, CardDescription, FloatIcons, Icon, CardInnerWrapper } from './Card.styles';

const Card = ({
  id,
  imageSrc,
  title,
  description,
  isShadowEnabled,
  onClickCard,
  showActionIcons,
  onClickDeleteIcon,
  onClickEditIcon
}) => {

  return (
    <CardWrapper data-testid={id} isShadowEnabled={isShadowEnabled}>
      {showActionIcons && (
        <FloatIcons>
          <Icon onClick={() => onClickDeleteIcon(id)} data-testid={`delete-ic-${id}`} >
            <DeleteIc/>
          </Icon>
          <Icon onClick={() => onClickEditIcon(id)} data-testid={`edit-ic-${id}`}>
            <EditIc />
          </Icon>
        </FloatIcons>
      )}
      <CardInnerWrapper data-testid={`card-content-${id}`} onClick={() => onClickCard(id)}>
        <CardImage src={imageSrc} alt={`card-image-${id}`} data-testid={`card-image-${id}`} />
        <CardContent>
          <CardTitle data-testid={`card-title-${id}`}>{title}</CardTitle>
          {description &&
            <CardDescription data-testid={`card-description-${id}`}>{description}</CardDescription>
          }
        </CardContent>
      </CardInnerWrapper>
    </CardWrapper>
  )
}

Card.propTypes = {
  id: PropTypes.number.isRequired,
  imageSrc: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  isShadowEnabled: PropTypes.bool,
  onClickCard: PropTypes.func,
  onClickDeleteIcon: PropTypes.func,
  onClickEditIcon: PropTypes.func,
  showActionIcons: PropTypes.bool,
}

Card.defaultProps = {
  isShadowEnabled: false,
  showActionIcons: false,
  description: null,
  onClickDeleteIcon: () => { },
  onClickEditIcon: () => {},
  onClickCard: () => { },
}

export default Card;