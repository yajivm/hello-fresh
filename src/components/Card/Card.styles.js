import styled from 'styled-components';

export const CardImage = styled.img`
  width: 100%;
  height: 250px;
  object-fit: cover;
`

export const CardContent = styled.div`
  padding: 10px 10px 20px;
`

export const CardTitle = styled.h3`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  margin: 5px 0 10px;
`

export const CardDescription = styled.div``

export const FloatIcons = styled.div`
  position: absolute;
  right: 10px;
  top: 10px;
  display: none;
`

export const Icon = styled.div`
  background: #fffefe75;
  border-radius: 30px;
  height: 25px;
  width: 30px;
  text-align: center;
  padding-top: 1px;
  transition: 0.3s;
  margin: 0 5px;
`

export const CardWrapper = styled.div`
  overflow: hidden;
  text-align: left;
  position: relative;
  ${({ isShadowEnabled }) => isShadowEnabled ?
    `
      box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px;
      background: #fff;
    `
  :
    `
      > img {
        border-radius: 8px;
      }
      > div {
        padding-left: 0;
      }
    `
  }

  &:hover > ${FloatIcons} {
    display: flex;
  }
`

export const CardInnerWrapper = styled.div``