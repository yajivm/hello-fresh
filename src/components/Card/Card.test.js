import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Card from '.'

const mockClickCard = jest.fn();
const mockClickDeleteIc = jest.fn();

const props = {
  id: 0,
  imageSrc: 'https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/garlic-herb-beef-burger-eeb7543e.jpg',
  title: 'Garlic & Herb Beef Burger with Fries',
  description: 'with Apple Slaw, BBQ Mayo & Almonds',
  onClickCard: mockClickCard,
}

describe('Card', () => {

  it('should render component', () => {
    const { container } = render(<Card {...props} />)

    expect(container).toMatchSnapshot();
  })

  it('should call onClick prop card when click card', () => {
    render(<Card {...props} />)

    fireEvent.click(screen.getByTestId('card-content-0'))

    expect(mockClickCard).toHaveBeenCalledTimes(1);
  })

  it('should show delete icon when showActionIcons prop is true', () => {
    const customProps = {
      ...props,
      showActionIcons: true,
      onClickDeleteIcon: mockClickDeleteIc
    }
    render(<Card {...customProps} />)

    expect(screen.getByTestId('delete-ic-0')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('delete-ic-0'))

    expect(mockClickDeleteIc).toHaveBeenCalledTimes(1);
  })

  it('should not show description text when description prop is false', () => {
    const customProps = {
      id: 0,
      imageSrc: 'https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/garlic-herb-beef-burger-eeb7543e.jpg',
      title: 'Garlic & Herb Beef Burger with Fries',
      onClickCard: mockClickCard,
    }
    render(<Card {...customProps} />)

    expect(screen.queryByTestId('card-description-0')).not.toBeInTheDocument();
  })
})