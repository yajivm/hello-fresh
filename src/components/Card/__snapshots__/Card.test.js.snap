// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render component 1`] = `
<div>
  <div
    class="sc-gKXOVf hdSzut"
    data-testid="0"
  >
    <div
      class="sc-iBkjds"
      data-testid="card-content-0"
    >
      <img
        alt="card-image-0"
        class="sc-bczRLJ blDScb"
        data-testid="card-image-0"
        src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/garlic-herb-beef-burger-eeb7543e.jpg"
      />
      <div
        class="sc-gsnTZi hhwjBW"
      >
        <h3
          class="sc-dkzDqf knkAIX"
          data-testid="card-title-0"
        >
          Garlic & Herb Beef Burger with Fries
        </h3>
        <div
          class="sc-hKMtZM"
          data-testid="card-description-0"
        >
          with Apple Slaw, BBQ Mayo & Almonds
        </div>
      </div>
    </div>
  </div>
</div>
`;
